sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExecBase.scala

```diff
@@ -646,26 +633,39 @@ abstract class GpuBroadcastNestedLoopJoinExecBase(
           // Everything else is treated like an unconditional cross join
           val buildSide = gpuBuildSide
           val joinTime = gpuLongMetric(JOIN_TIME)
+          val buildSchema = buildPlan.schema
+          val postProjection = buildSidePostProjection
           streamed.executeColumnar().mapPartitions { streamedIter =>
-            val lazyStream = streamedIter.map { cb =>
+            // Will materialize the stream batch before materializing the build batch.
+            val (buildBatch, bufferedStream) = makeBuiltBatchAndStreamIter(
+              relation,
+              streamedIter,
+              buildSchema,
+              buildTime,
+              buildDataSize)
+            val prjBuildBatch = postProjection.map(_(buildBatch)).getOrElse(buildBatch)
```

Comment:
Better to wrap this by `closeOnExcept` to avoid memory leak when any exception comes up during the project.
```
val prjBuildBatch = closeOnExcept(buildBatch) { _ =>
   postProjection.map(_(buildBatch)).getOrElse(buildBatch)
}
```